src/transformStateWithClones.js

```diff
@@ -7,7 +7,42 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let prevState = state;
+
+  for (const action of actions) {
+    let newState = transform (prevState, action);
+    result.push(newState);
+    prevState = newState;
+  }
+
+  return result;
+}
+
+function transform (prevState, action) {
+  switch (action.type) {
+    case 'clear':
+      return {};
+
+    case 'addProperties':
+      return {
+        ...prevState,
+        ...action.extraData,
+      };
+
+    case 'removeProperties': {
+      const newState = { ...prevState };
+
+      for (let keys of action.keysToRemove) {
+        delete newState[keys];
+      }
+
+      return newState;
+    }
+
+    default:
+      return prevState;
```

Comment:
you return the same object here, which means that if you'd change it, the previous value in array would change too, you should return the copy of it
```suggestion
      return { ...prevState };
```